storage/mysql/mysql.go

```diff
@@ -0,0 +1,294 @@
+	return func(c *config) {
+		c.driver = driver
+	}
+}
+
+// WithDB sets a custom MySQL *sql.DB to the storage.
+//
+// If set, driver passed via WithDriver is ignored.
+func WithDB(db *sql.DB) Option {
+	return func(c *config) {
+		c.db = db
+	}
+}
+
+// New creates and returns a new MySQLStorage.
+func New(opts ...Option) (*MySQLStorage, error) {
+	cfg := &config{driver: "mysql"}
+	for _, opt := range opts {
+		opt(cfg)
+	}
+	var err error
+	if cfg.db == nil {
+		cfg.db, err = sql.Open(cfg.driver, cfg.dsn)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if err = cfg.db.Ping(); err != nil {
+		return nil, err
+	}
+	return &MySQLStorage{db: cfg.db}, nil
+}
+
+// RetrieveAuthTokens reads the DEP OAuth tokens for name DEP name.
+func (s *MySQLStorage) RetrieveAuthTokens(ctx context.Context, name string) (*client.OAuth1Tokens, error) {
+	tokens := new(client.OAuth1Tokens)
+	err := s.db.QueryRowContext(
+		ctx, `
+SELECT
+    consumer_key,
+    consumer_secret,
+    access_token,
+    access_secret,
+	access_token_expiry
+FROM
+    dep_names
+WHERE
+    name = ?;`,
+		name,
+	).Scan(
+		&tokens.ConsumerKey,
+		&tokens.ConsumerSecret,
+		&tokens.AccessToken,
+		&tokens.AccessSecret,
+		&tokens.AccessTokenExpiry,
+	)
+	if err != nil {
+		return nil, err
+	}
+	return tokens, nil
+}
+
+// StoreAuthTokens saves the DEP OAuth tokens for the DEP name.
+func (s *MySQLStorage) StoreAuthTokens(ctx context.Context, name string, tokens *client.OAuth1Tokens) error {
+	_, err := s.db.ExecContext(
+		ctx, `
+INSERT INTO dep_names 
+	(name, consumer_key, consumer_secret, access_token, access_secret, access_token_expiry)
+VALUES 
+	(?, ?, ?, ?, ?, ?)
+ON DUPLICATE KEY UPDATE 
+	consumer_key = VALUES(consumer_key),
+	consumer_secret = VALUES(consumer_secret),
+	access_token = VALUES(access_token),
+	access_secret = VALUES(access_secret),
+	access_token_expiry = VALUES(access_token_expiry);`,
+		name,
+		tokens.ConsumerKey,
+		tokens.ConsumerSecret,
+		tokens.AccessToken,
+		tokens.AccessSecret,
+		tokens.AccessTokenExpiry,
+	)
+	return err
+}
+
+// RetrieveConfig reads the DEP config for the DEP name.
+//
+// Returns an empty config if the config does not exist (to support a fallback default config).
+func (s *MySQLStorage) RetrieveConfig(ctx context.Context, name string) (*client.Config, error) {
+	var baseURL sql.NullString
+	err := s.db.QueryRowContext(
+		ctx,
+		`SELECT config_base_url FROM dep_names WHERE name = ?;`,
+		name,
+	).Scan(
+		&baseURL,
+	)
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			// an 'empty' config is valid
+			return &client.Config{}, nil
+		}
+		return nil, err
+	}
+	var config client.Config
+	if baseURL.Valid {
+		config.BaseURL = baseURL.String
+	}
+	return &config, nil
+}
+
+// StoreConfig saves the DEP config for name DEP name.
+func (s *MySQLStorage) StoreConfig(ctx context.Context, name string, config *client.Config) error {
+	if config == nil || config.BaseURL == "" {
+		return nil
+	}
```

Comment:
I don't have much context here, so please lmk if you disagree. Should this be returning an error rather than silently doing nothing? Or, is it possible that it could be intentional to set the BaseURL to an empty string?

Reply:
API guard is done here: c77808e90a2965c6791a8de56a6cf61cea4c0b9b